rfcs/text/0015_nodejs_clustering.md

```diff
@@ -0,0 +1,554 @@
+as it doesn't actually run a Kibana server.
+
+There are various ways to address such use-cases. What seems to be the best compromise right now would be the concept of 
+'main worker'. The coordinator would arbitrary elect a worker as the 'main' one at startup. The clustering service would then expose 
+an API to let workers identify themselves as main or not.
+
+```ts
+export interface ClusteringServiceSetup {
+  // [...]
+  isMainWorker: () => boolean;
+}
+```
+
+Notes:
+- in non-clustered mode, `isMainWorker` would always return true, to reduce the divergence between clustered and 
+  non-clustered modes.
+
+## The clustering service API
+
+We will be adding a new clustering service to core, that will add the necessary cluster APIs, and would be accessible via core's setup and start contracts (`coreSetup.clustering` and `coreStart.clustering`).
+
+At the moment, no need to extend Core's request handler context with clustering related APIs has been identified.
+
+The contract interface would look like (more APIs could be added depending on the discussions on this RFC)
+
+```ts
+type ClusterMessagePayload = Serializable;
+
+interface BroadcastOptions {
+  /**
+   * If true, will also send the message to the worker that sent it.
+   * Defaults to false.
+   */
+  sendToSelf?: boolean;
+  /**
+   * If true, the message will also be sent to subscribers subscribing after the message was effectively sent.
+   * Defaults to false.
+   */
+  persist?: boolean;
+}
+
+export interface ClusteringServiceSetup {
+  /**
+   * Return true if clustering mode is enabled, false otherwise
+   */
+  isEnabled: () => boolean;
+  /**
+   * Return the current worker's id. In non-clustered mode, will return `1`
+   */
+  getWorkerId: () => number;
+  /**
+   * Broadcast a message to other workers.
+   * In non-clustered mode, this is a no-op
+   */
+  broadcast: (type: string, payload?: ClusterMessagePayload, options?: BroadcastOptions) => void;
+  /**
+   * Registered a handler for given `type` of IPC messages
+   * In non-clustered mode, this is a no-op that returns a no-op unsubscription callback.
+   */
+  addMessageHandler: (type: string, handler: MessageHandler) => MessageHandlerUnsubscribeFn;
+  /**
+   * Returns true if the current worker has been elected as the main one. In non-clustered mode, will always return true
+   */
+  isMainWorker: () => boolean;
+}
+```
+
+### Example: SO Migration
+
+To take the example of SO migration, the `KibanaMigrator.runMigrations` implementation could change to
+(naive implementation, the function is supposed to return a promise here, did not include that for simplicity):
+
+```ts
+runMigration() {
+   if(clustering.isMainWorker()) {
+     this.runMigrationsInternal().then((result) => {
+        applyMigrationState(result);
+        clustering.broadcast('migration-complete', { payload: result }, { persist: true }); // persist: true will send message even if subscriber subscribe after the message was actually sent
```

Comment:
> If we used Observables as the emit and subscribe interface, I don't believe we would need this persist option

It mostly depends on the underlying observable implementation we choose to implement. We need to keep in mind that if we use a observable pattern, we will need to code the implementation ourselves, as the actual IPC protocol we're using is event-based. So we'll be just creating an observable from the IPC messages exposed by the `clusteringService` on the workers side. 

If we get rid of this `persist` option and choose to use observable, and decide than a new subscriber should receive all previous events, we will then need to keep ALL previous IPC messages in memory, which could have some serious memory impact on a long-running Kibana instance.

>  For instance, if there is some sort of shared state that needs to be broadcast, it'd be nice to use a ReplaySubject(1)

This would not solve the synchronisation issues, as, unless we do store all messages to re-emit them, the `replay` effect would still have to be applied somewehere in the chain **before** the actual emission of the initial event go through said chain, which, in given example, is not necessarily true depending on processes race.  

So yea, we probably do want an Obs-based API, but that doesn't quite solve the technical challenges of pub/sub using IPC.

Reply:
I added observable-based API to the RFC as a future enhancement to investigate, as described in my comments below.